File: typescript/advanced/06-direct-protocol.ts
/**
 * Advanced API Example 06: Direct Protocol Access
 *
 * Demonstrates direct access to protocol modules (Kernel, Escrow, Events).
 * This is the lowest level API for maximum control.
 *
 * Available Protocol Modules:
 * - ACTPKernel: Transaction lifecycle management
 * - EscrowVault: Fund management
 * - EventMonitor: Blockchain event monitoring
 * - EASHelper: Attestation verification
 * - MessageSigner: EIP-712 message signing
 *
 * This example demonstrates:
 * - Accessing protocol modules via BlockchainRuntime
 * - Direct smart contract interaction
 * - Low-level protocol operations
 *
 * Note: This level is for advanced users building custom integrations.
 * Run: npx ts-node advanced/06-direct-protocol.ts
 */

import 'dotenv/config';
import {
  ACTPClient,
  BlockchainRuntime,
  ACTPKernel,
  EscrowVault,
  EventMonitor,
  getNetwork,
} from '@agirails/sdk';
import { log, logSection } from '../src/utils/helpers';

async function main() {
  logSection('AGIRAILS Advanced API - Direct Protocol Access');

  // =====================================================
  // Part 1: Protocol Module Overview
  // =====================================================
  logSection('Part 1: Protocol Modules');

  console.log('The SDK provides access to low-level protocol modules:');
  console.log('');
  console.log('┌─────────────────────────────────────────────────────────────┐');
  console.log('│  ACTPClient                                                 │');
  console.log('│  ├── beginner (BeginnerAdapter)                             │');
  console.log('│  ├── intermediate (IntermediateAdapter)                     │');
  console.log('│  └── advanced (IACTPRuntime)                                │');
  console.log('│       └── BlockchainRuntime (testnet/mainnet)               │');
  console.log('│            ├── getKernel() → ACTPKernel                     │');
  console.log('│            ├── getEscrow() → EscrowVault                    │');
  console.log('│            ├── getEvents() → EventMonitor                   │');
  console.log('│            ├── getEASHelper() → EASHelper                   │');
  console.log('│            └── getMessageSigner() → MessageSigner           │');
  console.log('└─────────────────────────────────────────────────────────────┘');
  console.log('');

  // =====================================================
  // Part 2: Network Configuration
  // =====================================================
  logSection('Part 2: Network Configuration');

  log('📋', 'Available network configurations...');

  const baseSepolia = getNetwork('base-sepolia');
  console.log('\nBase Sepolia (Testnet):');
  console.log(`  Chain ID: ${baseSepolia.chainId}`);
  console.log(`  RPC URL: ${baseSepolia.rpcUrl}`);
  console.log(`  Block Explorer: ${baseSepolia.blockExplorer}`);
  console.log('  Contracts:');
  console.log(`    ACTPKernel: ${baseSepolia.contracts.actpKernel}`);
  console.log(`    EscrowVault: ${baseSepolia.contracts.escrowVault}`);
  console.log(`    USDC: ${baseSepolia.contracts.usdc}`);
  console.log(`    EAS: ${baseSepolia.contracts.eas}`);

  // Base Mainnet - contracts not yet deployed
  console.log('\nBase Mainnet:');
  console.log('  Chain ID: 8453');
  console.log('  USDC: 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913');
  console.log('  (Kernel/Escrow: Not yet deployed - coming soon)');

  // =====================================================
  // Part 3: ACTPKernel Methods
  // =====================================================
  logSection('Part 3: ACTPKernel');

  console.log('ACTPKernel is the main transaction coordinator.');
  console.log('');
  console.log('Core Methods:');
  console.log('');
  console.log('  // Transaction creation');
  console.log('  kernel.createTransaction(params) → txId');
  console.log('');
  console.log('  // State transitions');
  console.log('  kernel.transitionState(txId, newState, proof?)');
  console.log('  kernel.linkEscrow(txId, escrowVault, escrowId)');
  console.log('');
  console.log('  // Escrow release');
  console.log('  kernel.releaseEscrow(txId, attestationUID?)');
  console.log('');
  console.log('  // Dispute handling');
  console.log('  kernel.raiseDispute(txId, reason)');
  console.log('  kernel.resolveDispute(txId, resolution)');
  console.log('');
  console.log('  // Queries');
  console.log('  kernel.getTransaction(txId) → Transaction');
  console.log('  kernel.getEconomicParams() → { platformFeeBps, ... }');
  console.log('  kernel.isPaused() → boolean');

  // =====================================================
  // Part 4: EscrowVault Methods
  // =====================================================
  logSection('Part 4: EscrowVault');

  console.log('EscrowVault holds funds during transactions.');
  console.log('');
  console.log('Core Methods:');
  console.log('');
  console.log('  // Token approval (required before linking)');
  console.log('  escrow.approveToken(tokenAddress, amount)');
  console.log('');
  console.log('  // Queries');
  console.log('  escrow.getEscrow(escrowId) → Escrow');
  console.log('  escrow.getRemaining(escrowId) → bigint');
  console.log('');
  console.log('Security Notes:');
  console.log('  - Funds are locked until DELIVERED + dispute window');
  console.log('  - Only Kernel can release/refund (onlyKernel modifier)');
  console.log('  - 7-day timelock for emergency withdrawals');

  // =====================================================
  // Part 5: EventMonitor
  // =====================================================
  logSection('Part 5: EventMonitor');

  console.log('EventMonitor tracks blockchain events in real-time.');
  console.log('');
  console.log('Event Types:');
  console.log('');
  console.log('  // Transaction events');
  console.log('  events.onTransactionCreated((txId, requester, provider) => { ... })');
  console.log('  events.onStateChanged((txId, oldState, newState) => { ... })');
  console.log('');
  console.log('  // Escrow events');
  console.log('  events.onEscrowLinked((txId, escrowId, amount) => { ... })');
  console.log('  events.onEscrowReleased((escrowId, recipient, amount) => { ... })');
  console.log('');
  console.log('  // Dispute events');
  console.log('  events.onDisputeRaised((txId, raiser, reason) => { ... })');
  console.log('  events.onDisputeResolved((txId, resolution) => { ... })');
  console.log('');
  console.log('  // Transaction watching');
  console.log('  events.watchTransaction(txId, callback)');
  console.log('  events.waitForState(txId, targetState, timeout)');

  // =====================================================
  // Part 6: Code Example (testnet)
  // =====================================================
  logSection('Part 6: Code Example');

  console.log('Direct protocol access in testnet/mainnet:');
  console.log('');
  console.log('```typescript');
  console.log("import { ACTPClient, BlockchainRuntime } from '@agirails/sdk';");
  console.log('import { ethers } from "ethers";');
  console.log('');
  console.log('// Create client with testnet config');
  console.log('const client = await ACTPClient.create({');
  console.log("  mode: 'testnet',");
  console.log('  requesterAddress: wallet.address,');
  console.log('  privateKey: wallet.privateKey,');
  console.log('  rpcUrl: process.env.RPC_URL,');
  console.log('});');
  console.log('');
  console.log('// Access BlockchainRuntime');
  console.log('const runtime = client.advanced as BlockchainRuntime;');
  console.log('');
  console.log('// Get protocol modules');
  console.log('const kernel = runtime.getKernel();');
  console.log('const escrow = runtime.getEscrow();');
  console.log('const events = runtime.getEvents();');
  console.log('');
  console.log('// Direct kernel interaction');
  console.log('const params = await kernel.getEconomicParams();');
  console.log('console.log("Platform fee:", params.platformFeeBps / 100, "%");');
  console.log('');
  console.log('// Set up event listener');
  console.log('events.onTransactionCreated((txId, requester, provider) => {');
  console.log('  console.log("New tx:", txId);');
  console.log('});');
  console.log('');
  console.log('// Create transaction via kernel');
  console.log('const txId = await kernel.createTransaction({');
  console.log('  provider: "0x...",');
  console.log('  requester: wallet.address,');
  console.log('  amount: ethers.parseUnits("100", 6), // 100 USDC');
  console.log('  deadline: Math.floor(Date.now() / 1000) + 3600,');
  console.log('  disputeWindow: 7200,');
  console.log('  serviceHash: ethers.keccak256(ethers.toUtf8Bytes("my-service")),');
  console.log('});');
  console.log('');
  console.log('// Approve USDC for escrow');
  console.log('await escrow.approveToken(usdcAddress, amount);');
  console.log('');
  console.log('// Link escrow');
  console.log('await kernel.linkEscrow(txId, escrowVaultAddress, escrowId);');
  console.log('```');

  // =====================================================
  // Part 7: When to Use Direct Access
  // =====================================================
  logSection('Part 7: When to Use Direct Access');

  console.log('Use direct protocol access when you need:');
  console.log('');
  console.log('✅ Custom gas management');
  console.log('   - Override default gas settings');
  console.log('   - Implement gas price oracles');
  console.log('');
  console.log('✅ Complex transaction patterns');
  console.log('   - Multi-step atomic operations');
  console.log('   - Custom state machine workflows');
  console.log('');
  console.log('✅ Event-driven architecture');
  console.log('   - Real-time blockchain monitoring');
  console.log('   - Custom event aggregation');
  console.log('');
  console.log('✅ Integration with other contracts');
  console.log('   - DEX integrations');
  console.log('   - Cross-protocol operations');
  console.log('');
  console.log('❌ Prefer higher-level APIs when:');
  console.log('   - Building simple integrations');
  console.log('   - Rapid prototyping');
  console.log('   - Standard transaction patterns');

  log('🎉', 'Direct protocol access demo complete!');
}

main().catch(console.error);

File: python/advanced/06_direct_protocol.py
#!/usr/bin/env python3
"""
Advanced API Example 06: Direct Protocol Access

Demonstrates direct access to ACTP protocol modules.
For advanced users who need full control over the protocol.

This example demonstrates:
- Direct runtime access
- Kernel module usage
- Escrow module usage
- Event monitoring module
- Time manipulation (mock mode)

Run: python advanced/06_direct_protocol.py
"""

import asyncio
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, log, log_section, format_state, format_usdc


async def main() -> None:
    log_section("AGIRAILS Advanced API - Direct Protocol Access")

    clear_mock_state()

    from agirails import (
        ACTPClient,
        MockRuntime,
        is_mock_runtime,
        State,
        CreateTransactionParams,
    )

    # =====================================================
    # Part 1: Runtime Architecture
    # =====================================================
    log_section("Part 1: Runtime Architecture")

    print("ACTP Protocol consists of several modules:")
    print()
    print("  IACTPRuntime (interface)")
    print("  ├── MockRuntime (in-memory, for testing)")
    print("  └── BlockchainRuntime (real blockchain)")
    print()
    print("  Runtime provides:")
    print("  ├── create_transaction()")
    print("  ├── link_escrow()")
    print("  ├── transition_state()")
    print("  ├── get_transaction()")
    print("  ├── get_balance()")
    print("  └── time (TimeInterface)")
    print()

    # =====================================================
    # Part 2: Direct Runtime Creation
    # =====================================================
    log_section("Part 2: Direct Runtime Creation")

    log("🔧", "Creating runtime directly (without ACTPClient)...")

    # Create mock runtime directly
    runtime = MockRuntime()
    await runtime.initialize()

    requester = "0x1111111111111111111111111111111111111111"
    provider = "0x2222222222222222222222222222222222222222"

    log("✅", "Runtime initialized")
    print(f"   Type: {type(runtime).__name__}")
    print(f"   Is Mock: {is_mock_runtime(runtime)}")
    print()

    # =====================================================
    # Part 3: Direct Transaction Creation
    # =====================================================
    log_section("Part 3: Direct Transaction Creation")

    log("📝", "Creating transaction with full parameters...")

    # Mint tokens first
    await runtime.mint_tokens(requester, 100_000_000)
    await runtime.mint_tokens(provider, 100_000_000)

    # Create transaction with all parameters
    tx_id = await runtime.create_transaction(CreateTransactionParams(
        provider=provider,
        requester=requester,
        amount=50_000_000,  # 50 USDC in wei
        deadline=int(time.time()) + 7200,  # 2 hours
        dispute_window=3600,  # 1 hour
        service_description='{"service": "direct-protocol-demo"}',
    ))

    print(f"   Transaction ID: {tx_id[:20]}...")

    # Get transaction details
    tx = await runtime.get_transaction(tx_id)
    print(f"   State: {format_state(tx.state)}")
    print(f"   Amount: {format_usdc(tx.amount)}")
    print(f"   Provider: {tx.provider[:16]}...")
    print(f"   Requester: {tx.requester[:16]}...")
    print()

    # =====================================================
    # Part 4: Direct Escrow Operations
    # =====================================================
    log_section("Part 4: Direct Escrow Operations")

    log("💳", "Linking escrow directly...")

    escrow_id = await runtime.link_escrow(tx_id, 50_000_000)
    print(f"   Escrow ID: {escrow_id[:20]}...")

    # Verify state changed to COMMITTED
    tx = await runtime.get_transaction(tx_id)
    print(f"   State: {format_state(tx.state)}")

    # Check balances
    requester_balance = await runtime.get_balance(requester)
    print(f"   Requester balance: {format_usdc(requester_balance)}")

    # =====================================================
    # Part 5: Direct State Transitions
    # =====================================================
    log_section("Part 5: Direct State Transitions")

    log("🔄", "Transitioning through states...")

    # COMMITTED -> IN_PROGRESS
    await runtime.transition_state(tx_id, State.IN_PROGRESS)
    tx = await runtime.get_transaction(tx_id)
    print(f"   State: {format_state(tx.state)}")

    # IN_PROGRESS -> DELIVERED
    await runtime.transition_state(tx_id, State.DELIVERED)
    tx = await runtime.get_transaction(tx_id)
    print(f"   State: {format_state(tx.state)}")

    # =====================================================
    # Part 6: Time Manipulation (Mock Only)
    # =====================================================
    log_section("Part 6: Time Manipulation")

    if is_mock_runtime(runtime):
        log("⏰", "Using MockRuntime time interface...")

        print(f"   Current time: {runtime.time.now()}")

        # Advance time
        await runtime.time.advance_time(1800)  # 30 minutes
        print(f"   After +30min: {runtime.time.now()}")

        # Advance past dispute window
        await runtime.time.advance_time(2000)  # +33 more minutes
        print(f"   After +33min: {runtime.time.now()}")

        # Set specific time
        specific_time = int(time.time()) + 86400  # Tomorrow
        runtime.time.set_time(specific_time)
        print(f"   Set to tomorrow: {runtime.time.now()}")

    else:
        print("   Time manipulation only available in mock mode")
    print()

    # =====================================================
    # Part 7: State Machine Rules
    # =====================================================
    log_section("Part 7: State Machine Rules")

    from agirails import is_valid_transition, is_terminal_state, STATE_TRANSITIONS

    print("Valid state transitions:")
    print()

    for state, valid_next in STATE_TRANSITIONS.items():
        next_states = ", ".join([format_state(s) for s in valid_next]) or "(none)"
        terminal = " [TERMINAL]" if is_terminal_state(state) else ""
        print(f"  {format_state(state)}{terminal}")
        print(f"    -> {next_states}")
        print()

    # =====================================================
    # Part 8: Error Handling
    # =====================================================
    log_section("Part 8: Error Handling")

    from agirails import (
        TransactionNotFoundError,
        InvalidStateTransitionError,
        InsufficientBalanceError,
    )

    log("⚠️", "Demonstrating protocol errors...")

    # Invalid transaction
    try:
        await runtime.get_transaction("0x" + "0" * 64)
        print("   (should have raised error)")
    except TransactionNotFoundError as e:
        print(f"   TransactionNotFoundError: {e.message[:40]}...")

    # Invalid state transition (can't go backwards)
    try:
        await runtime.transition_state(tx_id, State.COMMITTED)
    except InvalidStateTransitionError as e:
        print(f"   InvalidStateTransitionError: {e.message[:40]}...")

    print()

    # =====================================================
    # Part 9: Protocol Reset (Mock Only)
    # =====================================================
    log_section("Part 9: Protocol Reset")

    if is_mock_runtime(runtime):
        log("🔄", "Resetting mock runtime state...")

        # Count current state
        all_txs = await runtime.get_all_transactions()
        print(f"   Transactions before reset: {len(all_txs)}")

        # Reset
        await runtime.reset()

        all_txs = await runtime.get_all_transactions()
        print(f"   Transactions after reset: {len(all_txs)}")

    # =====================================================
    # Summary
    # =====================================================
    log_section("Summary")

    print("Direct protocol access provides:")
    print()
    print("  1. Full control over transaction lifecycle")
    print("  2. Direct escrow operations")
    print("  3. State machine manipulation")
    print("  4. Time control (mock mode)")
    print("  5. State reset (mock mode)")
    print()
    print("Use cases:")
    print("  - Custom transaction flows")
    print("  - Testing and simulation")
    print("  - Protocol-level integrations")
    print("  - Building alternative SDKs")

    log("🎉", "Direct protocol access demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
